### DISCUSSION TITLE
Automating release process

### DISCUSSION THREAD

**Discussion by ikedas** [sympa-community/sympa#1631] [Release works]
Discussion about automating release process.

### Resources

  * "[How to prepare a new source tarball](https://github.com/sympa-community/sympa/blob/sympa-6.2/support/README.support.md#how-to-prepare-a-new-source-tarball)" (`support/README.support.md` in source repo)

---

**Comment by ikedas** [comment]
I wrote a script that executes the steps 3 & 4 in above and pushes the changes.

It runs on the translation server and automatically pushes changes based on new translations using a "[deploy key](https://docs.github.com/developers/overview/managing-deploy-keys#deploy-keys)" registered on GitHub. It may be run periodically, e.g. once a week.

[Here](https://github.com/ikedas/sympa/compare/4c95474...0f5a959) is an example of the result of running it on my fork.

Changes to the translation server can be tampered with or mishandled, so they need to be checked by human eyes before being merged into the main branch (if a problem is found, we have to correct the translation on the translation server and wait for the next push by the automatic execution). That's why the new branch is created instead of direct commits on the main branch.

---
The point is that translation changes are always made on the translation server before being reflected in the repository. There are no changes in the reverse direction.

However, direction is reversed for translation templates (`.pot` files). Any changes to the source code have to be reflected on the translation server. I'm planning to make this part automatic as well.

**Comment by ikedas** [comment]
Now update of translation will be pushed **every Monday** into **the [`translation` branch in `sympa` repo](https://github.com/sympa-community/sympa/compare/translation)**.  We may freely merge it to "main" branch anytime we prefer.

**Comment by ikedas** [comment]
@racke, I could automate the steps 7 to 9 (making source distribution) using GitHub workflow:
  - [Result](https://github.com/ikedas/sympa/actions/runs/3279579607)
    The file "artifact.zip" in the bottom contains source distribution for 6.2.70.
  - [Workflow file](https://github.com/ikedas/sympa/blob/ec0172a4f6927e229962276f2171172c323a5fcd/.github/workflows/upload-dist.yml)

I'm not sure if these results are the same as what you are aiming for, but if this helps, please use it.

In above, the resulting source distribution files are stored in temporary storage, but they can also be uploaded to the release page automatically.   Please let me know if you want me to update the workflow file.

**Comment by racke** [comment]
Thanks, I will check that,

**Comment by ikedas** [comment]
### Automating release notes?
[Automatically generated release notes](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes) look enough for me ([animation](https://twitter.com/i/status/1585257089722908672)).

Currently we create a branch for release notes (e.g. `pr/release-6.2.70`) and manually transcribe to them every time we merge PRs, but such work seems excessive..

(Afaik the original intention of creating such a branch was to be able to know about noteworthy changes in advance.  However, since people complains about them after the new release without debugging them in advance anyway, I feel editing the release notes manually seems a waste of effort.  Really interested people will look at the PRs and milestones.)

**Comment by racke** [comment]
Looks good to me ... I will try that out ASAP.

**Comment by racke** [comment]
OK, I'm ready for the release.  Is NEWS.md up to date?

**Comment by ikedas** [comment]
It looks up-to-date for me, except the release date 😃

**Comment by racke** [comment]
Yes. will fix that.

**Comment by racke** [comment]
OK release prepared in GitHub. @ikedas please take a look. Will release it tomorrow.

**Comment by ikedas** [comment]
It looks good for me.

**Comment by racke** [comment]
Released on GitHub. How did you produce the message for the mailing list?

**Comment by ikedas** [comment]
In my case, I have copied the HTML
source of the release page, created the MIME  message using it by hand:smile: and fed it to sendmail(1).

**Comment by ikedas** [comment]
fyi: On the new mailing list server, the list for translators (ex. lets@translate.sympa.org) is included in the announcement list, so you may send annuncement only to `announce@sympa.community`.

**Comment by ikedas** [comment]
@racke, FYI I wrote a script to generate MIME message:
https://gist.github.com/ikedas/61b0b98814ce045b25e0438109cde00b
  * Note that the `From:`, `Date:` and `Message-ID:` header fields have to be added manually. 
  * Unfortunately it does not support embedded images.

**Comment by racke** [comment]
OK thanks will look at that as soon as possible.

**Comment by ikedas** [comment]
* Now steps 3 & 4 are automated

    Translation server periodically synchronizes `*.pot` file with the git repository, and pushes update of translations as the branch `translation`.  PR to merge this branch will be created as necessity: Recently https://github.com/sympa-community/sympa/pull/1519 has been created.

    In short, we may merge this PR into the main branch whenever. Then if there are more updates, the branch and PR will be created again.

  * @racke, could you please create `pr/release-x.x.x` branch for `NEWS.md` (or anything else you prefer)?  We need the place to write down "Incompatible changes".

**Comment by racke** [comment]
@ikedas done

**Comment by ikedas** [comment]
Thanks!

**Comment by racke** [comment]
> @racke, FYI I wrote a script to generate MIME message:
> https://gist.github.com/ikedas/61b0b98814ce045b25e0438109cde00b
> 
>     Note that the From:, Date: and Message-ID: header fields have to be added manually.
>     Unfortunately it does not support embedded images.

Unfortunately it barfs: open body: Invalid argument at /usr/share/perl5/MIME/Entity.pm line 1892.
Perl 5.32 installed from Debian 11 packages.

**Comment by ikedas** [comment]
What is the version of MIME-Tools ?

**Comment by ikedas** [comment]
@racke, please try this:
``` diff
$ diff -u support/github-release-message.pl.orig support/github-release-message.pl
--- support/github-release-message.pl.orig	2022-11-18 22:29:49.533269207 +0900
+++ support/github-release-message.pl	2022-11-18 22:28:03.677540670 +0900
@@ -24,6 +24,7 @@
 my $release  = decode_json($json);
 
 my $plain = sprintf "%s\n====\n\n%s", $release->{name}, $release->{body};
+Encode::_utf8_off($plain);
 my $html =
     sprintf
     "<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\"/>\n<title>%s</title>\n</head>\n<body>%s</body>\n</html>",
```

**Comment by racke** [comment]
I extended the script with SMTP and wrapping all links with <a href ....
Announce is out, but there is a glitch in the template:

To unsubscribe from this list, click here: /auto_signoff/announce?email=racke%40linuxia.de

I will clean up the script and send it to you.

**Comment by ikedas** [comment]
This seems the bug in Sympa.

Once you have improved the script to some extent, you can put it under /support/.

Anyway, **thank you so much for the release work! :tada:**

**Comment by racke** [comment]
> This seems the bug in Sympa.
> 

Can you open a bug report :-).

**Comment by ikedas** [comment]
Done!

**Comment by ikedas** [comment]
## Some ideas

*   Currently, release note text is generated only when the new release is tagged. Therefore, it is not possible to write a note (which cannot be generated automatically) about incompatible changes in the release notes.

    IMHO using [Release drafter](https://github.com/marketplace/actions/release-drafter) could solve this problem.

*   It haven't been decided whether to put the [ready] label on the issue or on the PR, but with the way we are working now, I think it would be better to put it on the PR (Similarly, [bug] and [enhancement] labels are put both on issues and PRs, but only those attached to PRs are used in the release notes).

**Comment by racke** [comment]
Release drafter sounds like a good idea. I also think that the the [ready] label should put on the PR.
